File: translator/src/providers/geminiProvider.ts
import type { TranslationSettings } from "../orchestrator/types";

type GeminiResponse = {
  candidates?: Array<{
    content?: {
      parts?: Array<{ text?: string }>;
    };
  }>;
};

export async function translateWithGemini(
  text: string,
  settings: TranslationSettings,
  signal: AbortSignal,
) {
  const prompt = [
    `Translate the following text into ${settings.targetLanguage}.`,
    "Preserve paragraph breaks, honorifics, names, tone, and formatting.",
    "Do not summarize, omit, add commentary, or wrap the result in markdown fences.",
    "",
    text,
  ].join("\n");

  const response = await fetch(
    `https://generativelanguage.googleapis.com/v1beta/models/${settings.model}:generateContent?key=${settings.apiKey}`,
    {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      signal,
      body: JSON.stringify({
        systemInstruction: {
          parts: [{ text: settings.systemPrompt }],
        },
        contents: [
          {
            role: "user",
            parts: [{ text: prompt }],
          },
        ],
      }),
    },
  );

  if (!response.ok) {
    throw new Error(await response.text());
  }

  const data = (await response.json()) as GeminiResponse;
  return data.candidates?.[0]?.content?.parts?.map((part) => part.text ?? "").join("") ?? "";
}

File: translator/python_app/providers/gemini_provider.py
from urllib import error, parse, request
import json

from python_app.config import TranslationSettings


def translate_with_gemini(text: str, settings: TranslationSettings) -> str:
    instructions = [
        f"Translate the following text into {settings.target_language}.",
        "Preserve paragraph breaks, honorifics, names, tone, and formatting.",
        "Do not summarize, omit, add commentary, or wrap the result in markdown fences.",
    ]
    if settings.translation_note.strip():
        instructions.extend(["", "Translation note:", settings.translation_note.strip()])
    if settings.glossary.strip():
        instructions.extend(["", "Glossary and name dictionary:", settings.glossary.strip()])

    prompt = "\n".join([*instructions, "", "Text to translate:", text])
    url = (
        "https://generativelanguage.googleapis.com/v1beta/models/"
        f"{parse.quote(settings.model)}:generateContent?key={parse.quote(settings.api_key)}"
    )
    payload = {
        "systemInstruction": {"parts": [{"text": settings.system_prompt}]},
        "contents": [{"role": "user", "parts": [{"text": prompt}]}],
    }
    body = json.dumps(payload).encode("utf-8")
    req = request.Request(url, data=body, method="POST", headers={"Content-Type": "application/json"})

    try:
        with request.urlopen(req, timeout=120) as response:
            data = json.loads(response.read().decode("utf-8"))
    except error.HTTPError as exc:
        detail = exc.read().decode("utf-8", errors="replace")
        raise RuntimeError(detail) from exc

    parts = data.get("candidates", [{}])[0].get("content", {}).get("parts", [])
    return "".join(part.get("text", "") for part in parts)
